File: fraiseql-typescript/examples/analytics_schema.ts
/**
 * Example FraiseQL analytics schema with fact tables.
 *
 * This example demonstrates:
 * - Fact table definitions with @FactTable()
 * - Aggregate query definitions with @AggregateQuery()
 * - Dimension paths and measure definitions
 * - Schema export with analytics support
 *
 * Usage:
 *   npx tsx examples/analytics_schema.ts
 *   # Creates schema.json with analytics fact tables
 */

import * as fraiseql from "../src/index";

// ============================================================================
// Type Definitions
// ============================================================================

/**
 * Sale event type with analytics dimensions.
 */
@fraiseql.type()
class Sale {
  id!: number;
  revenue!: number;
  quantity!: number;
  cost!: number;
  customerId!: string;
  occurredAt!: string;
}

// ============================================================================
// Field Registration
// ============================================================================

fraiseql.registerTypeFields("Sale", [
  { name: "id", type: "Int", nullable: false },
  { name: "revenue", type: "Float", nullable: false },
  { name: "quantity", type: "Int", nullable: false },
  { name: "cost", type: "Float", nullable: false },
  { name: "customerId", type: "String", nullable: false },
  { name: "occurredAt", type: "String", nullable: false },
]);

// ============================================================================
// Fact Table Definition
// ============================================================================

/**
 * Sales fact table for analytics.
 *
 * This fact table stores:
 * - Measures: revenue, quantity, cost (numeric aggregations)
 * - Dimensions: category, product_name (GROUP BY dimensions)
 * - Filters: customer_id, occurred_at (denormalized for fast WHERE)
 */
fraiseql.registerFactTableManual(
  "tf_sales",
  [
    { name: "revenue", sql_type: "Float", nullable: false },
    { name: "quantity", sql_type: "Int", nullable: false },
    { name: "cost", sql_type: "Float", nullable: false },
  ],
  {
    name: "data",
    paths: [
      {
        name: "category",
        json_path: "data->>'category'",
        data_type: "text",
      },
      {
        name: "product_name",
        json_path: "data->>'product_name'",
        data_type: "text",
      },
    ],
  },
  [
    { name: "customer_id", sql_type: "Text", indexed: true },
    { name: "occurred_at", sql_type: "Timestamp", indexed: true },
    { name: "id", sql_type: "Int", indexed: false },
  ]
);

// ============================================================================
// Aggregate Queries
// ============================================================================

/**
 * Sales aggregate query with flexible grouping and aggregation.
 *
 * Supports:
 * - groupBy: { category: true, occurred_at_day: true }
 * - aggregates: { count: true, revenue_sum: true, revenue_avg: true }
 * - where: { customer_id: { _eq: "uuid-123" } }
 * - having: { revenue_sum_gt: 1000 }
 * - orderBy: [{ field: "revenue_sum", direction: "DESC" }]
 * - limit: 100, offset: 0
 */
fraiseql.registerAggregateQuery(
  "salesAggregate",
  "tf_sales",
  true, // autoGroupBy
  true // autoAggregates
);

fraiseql.registerQuery(
  "salesAggregate",
  "Record",
  true, // returns list
  false, // not nullable
  [],
  "Aggregate sales data with flexible grouping and aggregation"
);

/**
 * Monthly sales trend query.
 *
 * Aggregates sales by month and category.
 */
fraiseql.registerAggregateQuery(
  "monthlySalesTrend",
  "tf_sales",
  true, // autoGroupBy
  true // autoAggregates
);

fraiseql.registerQuery(
  "monthlySalesTrend",
  "Record",
  true, // returns list
  false, // not nullable
  [{ name: "category", type: "String", nullable: true }],
  "Monthly sales trend by category"
);

/**
 * Customer revenue summary.
 *
 * Shows total and average revenue per customer.
 */
fraiseql.registerAggregateQuery(
  "customerRevenueSummary",
  "tf_sales",
  true, // autoGroupBy
  true // autoAggregates
);

fraiseql.registerQuery(
  "customerRevenueSummary",
  "Record",
  true, // returns list
  false, // not nullable
  [
    { name: "limit", type: "Int", nullable: true, default: 100 },
    { name: "minRevenue", type: "Float", nullable: true },
  ],
  "Top customers by revenue"
);

// ============================================================================
// Export Schema
// ============================================================================

// Export schema to JSON when run as main module
if (require.main === module) {
  fraiseql.exportSchema("schema.json");

  console.log("\n✅ Analytics schema exported successfully!");
  console.log("   Fact Tables: 1 (tf_sales)");
  console.log("   Aggregate Queries: 3");
  console.log("   ");
  console.log("   Next steps:");
  console.log("   1. Compile schema: fraiseql-cli compile schema.json");
  console.log("   2. Query sales data with aggregations:");
  console.log("      query {");
  console.log("        salesAggregate(");
  console.log("          groupBy: { category: true }");
  console.log("          aggregates: { count: true, revenue_sum: true }");
  console.log("        ) { ... }");
  console.log("      }");
}

// Also export for use as a module
export { Sale };

File: sdks/official/fraiseql-python/examples/analytics_schema.py
"""Example: Analytics schema with fact tables and aggregate queries.

This example demonstrates how to define fact tables for analytics workloads.
"""

import fraiseql


# Define a fact table type
@fraiseql.fact_table(
    table_name="tf_sales",
    measures=["revenue", "quantity", "cost"],
    dimension_paths=[
        {"name": "category", "json_path": "data->>'category'", "data_type": "text"},
        {"name": "product_name", "json_path": "data->>'product_name'", "data_type": "text"},
        {"name": "region", "json_path": "data->>'region'", "data_type": "text"},
    ],
)
@fraiseql.type
class Sale:
    """Sales transaction fact table.

    Fact table pattern:
    - Measures: revenue, quantity, cost (numeric columns for aggregation)
    - Dimensions: category, product_name, region (JSONB data column)
    - Denormalized filters: customer_id, product_id, occurred_at (indexed columns)
    """

    id: int
    # Measures (aggregatable)
    revenue: float
    quantity: int
    cost: float
    # Denormalized filters (indexed for fast WHERE)
    customer_id: str
    product_id: str
    occurred_at: str


# Define an aggregate query
@fraiseql.aggregate_query(
    fact_table="tf_sales",
    auto_group_by=True,
    auto_aggregates=True,
)
@fraiseql.query
def sales_aggregate() -> list[dict]:
    """Aggregate sales data with flexible grouping.

    Supports:
    - GROUP BY: category, product_name, region, occurred_at (day/week/month/year)
    - Aggregates: count, revenue_sum, revenue_avg, quantity_sum, cost_sum
    - WHERE: customer_id, product_id, occurred_at range
    - HAVING: revenue_sum_gt, quantity_sum_gte, etc.
    - ORDER BY: any aggregate or dimension
    - LIMIT/OFFSET: pagination

    Example GraphQL query:
    ```graphql
    query {
      sales_aggregate(
        where: {
          occurred_at: { _gte: "2025-01-01", _lt: "2025-02-01" }
          customer_id: { _eq: "customer-123" }
        }
        groupBy: {
          category: true
          occurred_at_day: true
        }
        having: {
          revenue_sum_gt: 1000
        }
        orderBy: [
          { field: "revenue_sum", direction: DESC }
        ]
        limit: 100
      ) {
        category
        occurred_at_day
        count
        revenue_sum
        revenue_avg
        quantity_sum
      }
    }
    ```
    """


# Regular types and queries can coexist
@fraiseql.type
class User:
    """User type."""

    id: int
    name: str
    email: str


@fraiseql.query(sql_source="v_user")
def users(limit: int = 10) -> list[User]:
    """Get all users."""


# Export schema
if __name__ == "__main__":
    fraiseql.export_schema("analytics_schema.json")
    print("✅ Analytics schema exported successfully!")
    print("   Fact tables: 1")
    print("   Aggregate queries: 1")
    print("   Regular types: 1")
    print("   Regular queries: 1")
